### DISCUSSION TITLE
generateStaticParams unexpected returns

### DISCUSSION THREAD

**Discussion by marc3601** [vercel/next.js#63376] [Help] [answered]
### Summary

I'm trying to generate slugs using `generateStaticParams` in Next but it somehow "chooses" diferent data to return that I want it to.

function `generateSlugs(params)` return array of slug objects each containing slug and id.

My problem is that `generateStaticParams` uses id as slug in returned params instead of slug. What could be wrong?

Here's my code from: produkt/[slug]/page.tsx

```typescript
import Gallery from "@/app/components/Gallery";
import ProductDescription from "@/app/components/ProductDescription";
import Navbar from "@/app/layout/Navbar";
import ProductSection from "@/app/layout/ProductSection";
import Products from "@/app/layout/Products";
import { Metadata, ResolvingMetadata } from "next";
import { notFound } from "next/navigation";
import { Slug } from "@/app/types/types";
import { ProductPage } from "@/app/types/types";
import { getData } from "@/app/utils/getData";
import generateSlugs from "@/app/utils/generateSlugs";

export const preferredRegion = ["fra1"];
export const dynamic = "force-dynamic";

export async function generateStaticParams() {
  const params = await fetch(`${process.env.API_URL}/api/auctions`).then(
    (res) => res.json()
  );

  const slugs = generateSlugs(params);

  return slugs.map((slug: Slug) => ({
    slug: slug.slug,
  }));
}

export async function generateMetadata(
  { params }: { params: Slug },
  parent: ResolvingMetadata
): Promise<Metadata> {
  console.log(params);

  const product = await getData(params);
  const url = "https://noanzo.pl";
  return {
    title: `${product[0]?.title} - noanzo.pl`,
    description: product[0]?.description,
    robots: "index, follow",
    alternates: {
      canonical: `${url}/produkt/${params.slug}`,
    },
    openGraph: {
      title: `${product[0]?.title} - noanzo.pl`,
      description: product[0]?.description,
      locale: "pl",
      images: [
        {
          url: (() => {
            const link = product[0]?.image.filter(
              (item: any) => item.thumbnail === true
            );
            if (!product[0]) {
              notFound();
            }
            if (link[0]?.url) {
              return link[0].url;
            } else {
              return product[0].image[0].url;
            }
          })(),
          alt: product[0]?.title,
        },
      ],
    },
  };
}

export default async function Page({ params }: { params: Slug }) {
  const data = await getData(params);
  if (!data || data[0] === null) {
    notFound();
  }
  const product: ProductPage = {
    title: data[0].title,
    description: data[0].description,
    price: data[0].price,
    image: data[0].image,
    imageLarge: data[0].imageLarge,
  };

  return (
    <main>
      <Navbar />
      <div className='container mx-auto sm:w-4/5 lg:w-11/12 xl:sm:w-4/5 p-2'>
        <ProductSection>
          <Gallery product={product} />
          <ProductDescription product={product} />
        </ProductSection>
      </div>
      <Products />
    </main>
  );
}
```

### Additional information

`generateStaticParams` return this: `{ slug: '9bc56d96-01e4-4bf9-b4a4-de7c31e6d544' }` and I want it to return something like this: `{slug: "buda-indywidualna-dla-psa-dc8d64de"}`

[Link to my api](https://admin.noanzo.pl/api/auctions)

**Comment by icyJoseph** [comment]
Show us the `generateSlugs` function.

**Reply by marc3601** [reply]
Here's generateSlugs:
```typescript
const generateSlugs = (params: Array<any>) => {
  const slugs = params.map((item: any) => {
    let slug = item.title.replace(/\s+/g, "-");
    if (!slug.endsWith("-")) {
      slug = slug + "-";
    }
    slug = slug + item.id.substring(0, 8);
    return { slug: slug.toLowerCase(), id: item.id };
  });
  return slugs;
};

export default generateSlugs;
```

**Reply by icyJoseph** [reply]
How are you seeing this issue exactly? It works fine for me, in latest Next.js at least.

Could it be that there are other links pointing to this page? that are not using the slug?

**Reply by marc3601** [reply]
It actually works fine and provides links like this: `http://localhost:3000/produkt/b533dd7a-c1b2-4201-8ccd-bbde1d7900ec`. The problem is that I need something like this instead: `http://localhost:3000/produkt/buda-z-tarasem-200-cm-7f8e5422`.
`generateSlugs` returns objects with an id and a slug, but `generateStaticParams` somehow "decides" to use the id instead of the slug. 
I have no clue what might be wrong and I probably messed up something really bad. :)

**Reply by icyJoseph** [reply]
Could you try to create a reproduction repository, or a Stackblitz example, with the minimal changes that trigger the issue? For Stackblitz just type next.new in your browser search bar.

**Reply by marc3601** [reply]
Hope it works: https://stackblitz.com/~/github.com/marc3601/noanzo-2.0
Branch "textEdit" latest commit
In `app/produkt/[slug]/page.tsx`  line 68 I log params that you can see while navigating to various "produkt" pages. This is how website works but now I need to change that so it has nice slugs instead of uuid.

To change it I created `generateSlugs` and I expect `generateStaticParams` to return slugs but instead I still see uuids. That's the confusing part for me. I'm new to Next so I'm sure I've missed something obvious. :)

**Reply by icyJoseph** [reply, marked as answer]
Hi,

The problem is not the `generateStaticParams` function at all. The error is here:

```tsx
  const thumbnails = shuffled.map((item: Auction, id: string): BadgeData => {
    let thumb = item.image.filter((item) => item.thumbnail === true);
    if (thumb.length === 0) {
      thumb = [item.image[0]];
    }

    return { title: item.title, link: thumb[0].url, slug: item.id };
  });
```

You return `slug: item.id` at the end, which is why the Links have that information. Also that `map` signature is wrong.

```tsx
        <Link key={i} href={`/produkt/${data.slug}`}> {/* here slug is item.id as per the mapping over shuffled */}
          <Badge data={data} />
        </Link>
```

So you have to also slugify these when generating the thumbnails, so that the links point to the correct product page. 

I think you'd have noticed this earlier, if you used: https://nextjs.org/docs/app/api-reference/functions/generate-static-params#generate-only-a-subset-of-params - That'd show a 404, when navigating from home to the product pages, because the links are incorrectly generated.

So you need to do:

```tsx
  const thumbnails = shuffled.map((item: Auction): BadgeData => {
    let thumb = item.image.filter((item) => item.thumbnail === true);
    if (thumb.length === 0) {
      thumb = [item.image[0]];
    }

    return { title: item.title, link: thumb[0].url, slug: generateSlug(item).slug };
  });
```

Where `generateSlug` is:

```ts
// one slug
export const generateSlug =  (item: any) => {
    let slug = item.title.replace(/\s+/g, "-");
    if (!slug.endsWith("-")) {
      slug = slug + "-";
    }
    slug = slug + item.id.substring(0, 8);
    return { slug: slug.toLowerCase(), id: item.id };
}

// many slugs
const generateSlugs = (params: Array<any>) => {
  const slugs = params.map(generateSlug);
  return slugs;
};

export default generateSlugs;
```

And that'll start to show the right parameters. Please do read more about https://nextjs.org/docs/app/api-reference/functions/generate-static-params and what does it actually do. Have a good day!

**Reply by marc3601** [reply]
Hi,

Yes it works! Thanks a lot. I really didn't understand how generateStaticParams work.
